pkg/template/controller/readiness.go

```diff
@@ -0,0 +1,139 @@
+
+	buildapi "github.com/openshift/origin/pkg/build/apis/build"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	"github.com/openshift/origin/pkg/client"
+	deployapi "github.com/openshift/origin/pkg/deploy/apis/apps"
+)
+
+func checkBuildReadiness(oc client.Interface, obj runtime.Object) (bool, bool, error) {
+	b := obj.(*buildapi.Build)
+
+	ready := buildutil.IsTerminalPhase(b.Status.Phase) && b.Status.Phase == buildapi.BuildPhaseComplete
+
+	failed := buildutil.IsTerminalPhase(b.Status.Phase) && b.Status.Phase != buildapi.BuildPhaseComplete
+
+	return ready, failed, nil
+}
+
+// TODO: this should be reported on the BuildConfig object itself.
+func checkBuildConfigReadiness(oc client.Interface, obj runtime.Object) (bool, bool, error) {
+	bc := obj.(*buildapi.BuildConfig)
+
+	builds, err := oc.Builds(bc.Namespace).List(metav1.ListOptions{LabelSelector: buildutil.BuildConfigSelector(bc.Name).String()})
+	if err != nil {
+		return false, false, err
+	}
+
+	for _, build := range builds.Items {
+		if build.Annotations[buildapi.BuildNumberAnnotation] == strconv.FormatInt(bc.Status.LastVersion, 10) {
+			return checkBuildReadiness(oc, &build)
+		}
+	}
+
+	return false, false, nil
+}
+
+func checkDeploymentReadiness(oc client.Interface, obj runtime.Object) (bool, bool, error) {
+	d := obj.(*extensions.Deployment)
+
+	var progressing, available *extensions.DeploymentCondition
+	for i, condition := range d.Status.Conditions {
+		switch condition.Type {
+		case extensions.DeploymentProgressing:
+			progressing = &d.Status.Conditions[i]
+
+		case extensions.DeploymentAvailable:
+			available = &d.Status.Conditions[i]
+		}
+	}
+
+	ready := progressing != nil &&
+		progressing.Status == kapi.ConditionTrue &&
+		progressing.Reason == deploymentutil.NewRSAvailableReason &&
+		available != nil &&
+		available.Status == kapi.ConditionTrue
+
+	failed := progressing != nil && progressing.Status == kapi.ConditionFalse
+
+	return ready, failed, nil
+}
+
+func checkDeploymentConfigReadiness(oc client.Interface, obj runtime.Object) (bool, bool, error) {
+	dc := obj.(*deployapi.DeploymentConfig)
+
+	var progressing, available *deployapi.DeploymentCondition
+	for i, condition := range dc.Status.Conditions {
+		switch condition.Type {
+		case deployapi.DeploymentProgressing:
+			progressing = &dc.Status.Conditions[i]
+
+		case deployapi.DeploymentAvailable:
+			available = &dc.Status.Conditions[i]
+		}
+	}
+
+	ready := progressing != nil &&
+		progressing.Status == kapi.ConditionTrue &&
+		progressing.Reason == deployapi.NewRcAvailableReason &&
+		available != nil &&
+		available.Status == kapi.ConditionTrue
+
+	failed := progressing != nil && progressing.Status == kapi.ConditionFalse
+
+	return ready, failed, nil
+}
+
+func checkStatefulSetReadiness(oc client.Interface, obj runtime.Object) (bool, bool, error) {
+	ss := obj.(*apps.StatefulSet)
+
+	ready := ss.Status.ReadyReplicas == ss.Spec.Replicas
```

Comment:
I would also check that metadata.generation equals to status.observedGeneration. In all cases and not just with statefulsets.